### PR TITLE
feat(lib/babe): implement secondary vrf slot block production

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes

- implement secondary vrf slot block production
- Also merge slot claim logic

## Tests

- Run a cross-client devnet with 2 nodes (with polkadot 0.9.10), https://hackmd.io/LfRK6aG6T7mMCE4Uybr-Mg. You will have to edit allowed slots in genesis-spec.json file to `"allowed_slots": "PrimaryAndSecondaryVRFSlots"`. Use the modifies genesis-spec.json to create genesis.json
- You will see blocks getting imported and built. On polkadot side, you should see messages suggesting that secondary vrf slots are being verified correctly. Something like below
```
2022-02-18 16:04:26.807 DEBUG tokio-runtime-worker babe: Verifying secondary VRF block #18 at slot: 274196743    
2022-02-18 16:04:26.807 TRACE             async-io async_io::driver: main_loop: waiting on I/O    
2022-02-18 16:04:26.807 TRACE             async-io async_io::reactor: process_timers: 0 ready wakers    
2022-02-18 16:04:26.807 TRACE             async-io polling: Poller::wait(_, Some(230.265526807s))    
2022-02-18 16:04:26.807 TRACE             async-io polling::epoll: wait: epoll_fd=22, timeout=Some(230.265526807s)    
2022-02-18 16:04:26.807 TRACE             async-io polling::epoll: modify: epoll_fd=22, fd=24, ev=Event { key: 18446744073709551615, readable: true, writable: false }    
2022-02-18 16:04:26.807 TRACE tokio-runtime-worker trie: Committing trie changes to db.    
```

- https://github.com/ChainSafe/gossamer/issues/2216

## Primary Reviewer

- @noot

### CODE REVIEW

**Comment by kishansagathiya** [comment]
Finally saw that magical log message suggesting that secondary vrf slots are being verified correctly
```
2022-02-18 16:04:34.092 TRACE tokio-runtime-worker babe: Verifying origin: NetworkBroadcast header: Header { parent_hash: 0x234028f29838f9e2251ed4ad046577b0da5ba4f23f765cc62bf68958c14a55c0, number: 19, state_root: 0xf365086251b5add92c26bc6ddbd7cb689ef436610c436374d67e655213856dbb, extrinsics_root: 0xf384a65025bb41419c7a3161c18e9d1998c1985528c0203cd3c04239bfaa01a5, digest: Digest { logs: [DigestItem::PreRuntime([66, 65, 66, 69], [3, 1, 0, 0, 0, 9, 233, 87, 16, 0, 0, 0, 0, 168, 200, 165, 140, 114, 229, 72, 133, 197, 72, 124, 118, 89, 74, 60, 215, 140, 76, 254, 164, 188, 104, 17, 182, 28, 237, 223, 63, 102, 47, 121, 37, 107, 33, 207, 158, 68, 68, 115, 23, 150, 182, 147, 83, 232, 127, 196, 158, 175, 36, 105, 206, 112, 197, 208, 90, 165, 212, 203, 138, 14, 65, 1, 12, 101, 40, 57, 139, 67, 35, 222, 11, 144, 26, 141, 23, 99, 233, 4, 171, 173, 106, 44, 115, 21, 70, 59, 173, 41, 142, 58, 229, 201, 81, 243, 8]), DigestItem::Seal([66, 65, 66, 69], [100, 222, 149, 189, 250, 93, 86, 50, 253, 153, 162, 172, 145, 230, 120, 198, 156, 188, 48, 79, 227, 97, 228, 150, 234, 152, 225, 99, 95, 197, 64, 19, 222, 163, 193, 219, 172, 28, 62, 237, 112, 220, 182, 7, 139, 143, 70, 12, 125, 254, 81, 90, 51, 187, 100, 165, 64, 54, 178, 133, 121, 3, 127, 131])] } } justification(s): None body: Some([0403000b09c3650c7f01])    
2022-02-18 16:04:34.092 DEBUG tokio-runtime-worker babe: We have 2 logs in this header    
2022-02-18 16:04:34.092 TRACE tokio-runtime-worker babe: Checking log DigestItem::PreRuntime([66, 65, 66, 69], [3, 1, 0, 0, 0, 9, 233, 87, 16, 0, 0, 0, 0, 168, 200, 165, 140, 114, 229, 72, 133, 197, 72, 124, 118, 89, 74, 60, 215, 140, 76, 254, 164, 188, 104, 17, 182, 28, 237, 223, 63, 102, 47, 121, 37, 107, 33, 207, 158, 68, 68, 115, 23, 150, 182, 147, 83, 232, 127, 196, 158, 175, 36, 105, 206, 112, 197, 208, 90, 165, 212, 203, 138, 14, 65, 1, 12, 101, 40, 57, 139, 67, 35, 222, 11, 144, 26, 141, 23, 99, 233, 4, 171, 173, 106, 44, 115, 21, 70, 59, 173, 41, 142, 58, 229, 201, 81, 243, 8]), looking for pre runtime digest    
2022-02-18 16:04:34.092 TRACE tokio-runtime-worker babe: Checking log DigestItem::Seal([66, 65, 66, 69], [100, 222, 149, 189, 250, 93, 86, 50, 253, 153, 162, 172, 145, 230, 120, 198, 156, 188, 48, 79, 227, 97, 228, 150, 234, 152, 225, 99, 95, 197, 64, 19, 222, 163, 193, 219, 172, 28, 62, 237, 112, 220, 182, 7, 139, 143, 70, 12, 125, 254, 81, 90, 51, 187, 100, 165, 64, 54, 178, 133, 121, 3, 127, 131]), looking for pre runtime digest    
2022-02-18 16:04:34.092 TRACE tokio-runtime-worker babe: Ignoring digest not meant for us    
2022-02-18 16:04:34.092 TRACE tokio-runtime-worker babe: Checking header    
2022-02-18 16:04:34.092 DEBUG tokio-runtime-worker header: Retrieving mutable reference to digest    
2022-02-18 16:04:34.092 DEBUG tokio-runtime-worker babe: Verifying secondary VRF block #19 at slot: 274196745    
2022-02-18 16:04:34.092 TRACE             async-io async_io::driver: main_loop: waiting on I/O    
2022-02-18 16:04:34.092 TRACE             async-io async_io::reactor: process_timers: 0 ready wakers    
2022-02-18 16:04:34.092 TRACE             async-io polling: Poller::wait(_, Some(222.980433576s))    
2022-02-18 16:04:34.092 TRACE             async-io polling::epoll: wait: epoll_fd=22, timeout=Some(222.980433576s)    
2022-02-18 16:04:34.092 TRACE             async-io polling::epoll: modify: epoll_fd=22, fd=24, ev=Event { key: 18446744073709551615, readable: true, writable: false }    
2022-02-18 16:04:34.093 TRACE tokio-runtime-worker trie: Committing trie changes to db.    
```

**Comment by noot** [comment]
@kishansagathiya what version of polkadot are you using to test this? on my machine it doesn't look like polkadot is able to import gossamer's blocks

File: lib/babe/crypto.go [hunk 1]
```diff
@@ -84,6 +84,37 @@ func checkPrimaryThreshold(randomness Randomness,
 	return inoutUint.Compare(threshold) < 0, nil
 }
 
+func claimSecondarySlotVRF(randomness Randomness,
```
Comment (noot) [inline lib/babe/crypto.go:87]: this is interesting, the VRF output isn't compared to any sort of threshold? I'm wondering what the point of the VRF is.. could you link to the substrate code? just wondering if they have any explanation of this

**Review by noot — COMMENTED** [review]
nice, looks very clean! let me know the versions you used to test so i can try it out

**Comment by kishansagathiya** [comment]
> @kishansagathiya what version of polkadot are you using to test this?

0.9.10 works. So, you can follow the same steps described in https://hackmd.io/LfRK6aG6T7mMCE4Uybr-Mg.

> on my machine it doesn't look like polkadot is able to import gossamer's blocks

Check if gossamer got disconnected. Looked for `dropped` in polkadot logs. If so, you you will have to restart both nodes.

File: dot/types/babe_digest.go [hunk 2]
```diff
@@ -119,5 +119,19 @@ func NewBabeSecondaryVRFPreDigest(authorityIndex uint32,
 	}
 }
 
+// ToPreRuntimeDigest returns the BabeSecondaryVRFPreDigest as a PreRuntimeDigest
+func (d *BabeSecondaryVRFPreDigest) ToPreRuntimeDigest() (*PreRuntimeDigest, error) {
+	digest := NewBabeDigest()
+	err := digest.Set(*d)
+	if err != nil {
+		return nil, err
```
Comment (qdm12) [inline dot/types/babe_digest.go:127]:
```suggestion
		return nil, fmt.Errorf("cannot set Babe secondary VRF pre-digest to Babe digest: %w", err)
```

File: dot/types/babe_digest.go [hunk 3]
```diff
@@ -119,5 +119,19 @@ func NewBabeSecondaryVRFPreDigest(authorityIndex uint32,
...
+	}
+	enc, err := scale.Marshal(digest)
+	if err != nil {
+		return nil, err
```
Comment (qdm12) [inline dot/types/babe_digest.go:131]:
```suggestion
		return nil, fmt.Errorf("cannot marshal Babe digest: %w", err)
```

File: dot/types/babe_digest.go [hunk 4]
```diff
@@ -119,5 +119,19 @@ func NewBabeSecondaryVRFPreDigest(authorityIndex uint32,
...
 }
 
+// ToPreRuntimeDigest returns the BabeSecondaryVRFPreDigest as a PreRuntimeDigest
+func (d *BabeSecondaryVRFPreDigest) ToPreRuntimeDigest() (*PreRuntimeDigest, error) {
+	digest := NewBabeDigest()
+	err := digest.Set(*d)
+	if err != nil {
+		return nil, err
+	}
+	enc, err := scale.Marshal(digest)
```
Comment (qdm12) [inline dot/types/babe_digest.go:129]: nit `babeDigestEncoding` would be nice. I'm an explicit/full words maniac so let me know if i'm bothering as well 😅

File: lib/babe/crypto.go [hunk 5]
```diff
@@ -84,6 +84,37 @@ func checkPrimaryThreshold(randomness Randomness,
...
+	slot, epoch uint64,
+	authorities []types.Authority,
+	keypair *sr25519.Keypair,
+	authorityIndex uint32,
+) (*VrfOutputAndProof, error) {
+
```
Comment (qdm12) [inline lib/babe/crypto.go:93]:
```suggestion
```

File: lib/babe/epoch.go [hunk 6]
```diff
@@ -221,37 +222,69 @@ func (b *Service) incrementEpoch() (uint64, error) {
...
+			return nil, fmt.Errorf(
+				"failed to get preruntime digest from babe secondary plain predigest for slot %d: %w", slotNumber, err)
 		}
-		return types.NewBabeSecondaryPlainPreDigest(epochData.authorityIndex, slot).ToPreRuntimeDigest()
+		logger.Debugf("epoch %d: claimed secondary plain slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
 	}
 
-	return nil, err
+	return nil, errors.New("invalid slot claiming technique")
```
Comment (qdm12) [inline lib/babe/epoch.go:289]: define a sentinel error

File: lib/babe/epoch.go [hunk 7]
```diff
@@ -221,37 +222,69 @@ func (b *Service) incrementEpoch() (uint64, error) {
...
 		epochData.threshold,
-		b.keypair,
+		keypair,
 	)
-	if err == nil {
-		return types.NewBabePrimaryPreDigest(
-			epochData.authorityIndex, slot, proof.output, proof.proof).ToPreRuntimeDigest()
+
+	switch err {
+	case nil:
```
Comment (qdm12) [inline lib/babe/epoch.go:242]: change this switch to an if / else if, no point having a `switch` here really

File: lib/babe/crypto.go [hunk 8]
```diff
@@ -84,6 +84,37 @@ func checkPrimaryThreshold(randomness Randomness,
...
+		logger.Debugf("claimed secondary slot, for slot number: %d", slot)
+
+		return &VrfOutputAndProof{
+			output: out,
+			proof:  proof,
+		}, nil
+	}
+
+	return nil, errNotOurTurnToPropose
```
Comment (qdm12) [inline lib/babe/crypto.go:115]: nit flip the condition around to be `if authorityIndex != secondarySlotAuthor` to return early and have less indentation

File: lib/babe/crypto.go (hunk 1 above)
Comment (noot) [reply lib/babe/crypto.go:87]: @kishansagathiya not something that needs to be changed code-wise, but just curious!

**Review by noot — APPROVED** [review]
tried on my home machine with v0.9.10 and it works! nice :D

File: lib/babe/epoch.go (hunk 6 above)
Comment (kishansagathiya) [reply lib/babe/epoch.go]: do you mean defining a variable for this? This error should never occur, I think we don't have to

File: lib/babe/crypto.go (hunk 1 above)
Comment (kishansagathiya) [reply lib/babe/crypto.go:87]: It seems like this vrf can be consumed by parachains, I am not sure why. https://github.com/paritytech/substrate/blob/master/client/consensus/babe/README.md

Comment (kishansagathiya) [reply lib/babe/crypto.go:87]: This is the relevant code 
https://github.com/paritytech/substrate/blob/c22fce5a311beede13479c9a00cca85d823b6b00/client/consensus/babe/src/authorship.rs#L158-L166

**Pull request merged**